### PR TITLE
ZCS-5101: create interface for email and sms in forget password functionality

### PR DESCRIPTION
**Pull request by chinmay15**
**Problem:** make code compatible, to easily add sms channel in future for account recovery.

**Fix:**
- create enum for Channel
- - email
- create interface for SendRecoveryCode
- - Implement EmailRecoveryCode(actual implementation to send emails)
- add param to RecoverAccountRequest
- - channel=“email” (this should be channel enum)
- - for other channels, invalid input
- change SetRecoveryEmailRequest -> SetRecoveryAccountRequest
- - add param channel=“email” (this should be channel enum)
- - rename recoveryEmailAddress -> recoveryAccount
- - rename recoveryEmailAddressVerificationCode -> recoveryAccountVerificationCode
- - this should send email using EmailRecoveryCode class

**Testing Done:**
- Added unit tests in RecoverAccountTest and SetRecoveryEmailTest.
- Corrected all other unit tests in the same.

**Testing to be done by QA:**
- Make changes in the existing automation to accommodate API changes.
- Test password reset functionality overall as per current implementation.

### CODE REVIEW

**Comment by chinmay15** [comment]
@shrikant-prasad has suggested some changes. Will do those and add in the PR.

File: soap/src/java/com/zimbra/soap/mail/message/RecoverAccountRequest.java [hunk 1]
```diff
@@ -46,12 +50,19 @@
     @XmlAttribute(name = MailConstants.A_EMAIL /* email */, required = true)
     private String email;
 
+    /**
+     * @zm-api-field-description email
```
Comment (desouzas) [inline soap/src/java/com/zimbra/soap/mail/message/RecoverAccountRequest.java]: `channel`

File: soap/src/java/com/zimbra/soap/mail/message/RecoverAccountResponse.java [hunk 2]
```diff
@@ -35,32 +35,32 @@
     /**
      * @zm-api-field-description RecoveryEmail
```
Comment (desouzas) [inline soap/src/java/com/zimbra/soap/mail/message/RecoverAccountResponse.java]: `RecoveryAccount` - optionally

File: soap/src/java/com/zimbra/soap/mail/message/RecoverAccountResponse.java [hunk 3]
```diff
@@ -35,32 +35,32 @@
...
     }
 
-    public RecoverAccountResponse(String recoveryEmail) {
-        this.recoveryEmail = recoveryEmail;
+    public RecoverAccountResponse(String recoveryAccount) {
+        this.recoveryAccount = recoveryAccount;
     }
 
     /**
      * @return the recovery recoveryEmail
```
Comment (desouzas) [inline soap/src/java/com/zimbra/soap/mail/message/RecoverAccountResponse.java]: `recoveryAccount` - optionally suppose it doesn't matter

File: soap/src/java/com/zimbra/soap/mail/message/SetRecoveryEmailRequest.java [hunk 4]
```diff
@@ -55,15 +56,22 @@ public static Op fromString(String s) throws ServiceException {
      * @zm-api-field-tag recoveryEmailAddress
```
Comment (desouzas) [inline soap/src/java/com/zimbra/soap/mail/message/SetRecoveryEmailRequest.java]: `recoveryAccount` ? optionally change desc. too

File: soap/src/java/com/zimbra/soap/mail/message/SetRecoveryEmailRequest.java [hunk 5]
```diff
@@ -55,15 +56,22 @@ public static Op fromString(String s) throws ServiceException {
...
      * @zm-api-field-description recovery email address
      */
-    @XmlAttribute(name = MailConstants.A_RECOVERY_EMAIL_ADDRESS /* recoveryEmailAddress */, required = false)
-    private String recoveryEmailAddress;
+    @XmlAttribute(name = MailConstants.A_RECOVERY_ACCOUNT /* recoveryAccount */, required = false)
+    private String recoveryAccount;
 
     /**
      * @zm-api-field-tag recoveryEmailAddressVerificationCode
```
Comment (desouzas) [inline soap/src/java/com/zimbra/soap/mail/message/SetRecoveryEmailRequest.java]: `recoveryAccountVerificationCode` ? Optionally update desc too.

File: store/docs/soap.txt [hunk 6]
```diff
@@ -4788,14 +4789,15 @@ reset: reads no input parameter. 'reset' will clear all ldap attributes related
...
 
-getRecoveryEmail = get recovery email only if it's verified.
+getRecoveryAccount = get recovery email only if it's verified.
 sendRecoveryCode = send recovery code to recovery email only if it's verified.
 user_email = email address of the user who forgot the password.
+channel = Mandatory attribute. Currently only email is supported.
 
-<RecoverAccountResponse recoveryEmail="{recory_email_address}" />
-recory_email_address = verified recovery email address for getRecoveryEmail operation
+<RecoverAccountResponse recoveryAccount="{recory_email_address}" />
```
Comment (desouzas) [inline store/docs/soap.txt]: `recory_email_address` typo

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java [hunk 7]
```diff
@@ -81,104 +68,90 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
...
+                                } else {
+                                    recoveryCodeMap.put(CodeConstants.RESEND_COUNT.toString(), String.valueOf(resendCount));
+                                }
+                            }
                         } else {
-                            recoveryCodeMap.put(CodeConstants.RESEND_COUNT.toString(), String.valueOf(resendCount));
+                            ZimbraLog.account.debug("%s Recovery code not found for %s, creating new one", LOG_OPERATION,
+                                    email);
+                            recoveryCodeMap.put(CodeConstants.EMAIL.toString(), recoveryEmail);
+                            recoveryCodeMap.put(CodeConstants.CODE.toString(), RandomStringUtils.random(8, true, true));
```
Comment (desouzas) [inline store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: Do we want a constant or comment for that `8`?

File: store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java [hunk 8]
```diff
@@ -81,104 +68,90 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
...
+                            ZimbraLog.account.debug("%s Recovery code found for %s", LOG_OPERATION, email);
+                            recoveryCodeMap = JWEUtil.getDecodedJWE(storedCodeString);
+                            ZonedDateTime storedDate = Instant
+                                    .ofEpochMilli(Long.valueOf(recoveryCodeMap.get(CodeConstants.EXPIRY_TIME.toString())))
+                                    .atZone(ZoneId.systemDefault());
+                            if (ChronoUnit.MILLIS.between(currentDate, storedDate) <= 0L) {
+                                ZimbraLog.account.debug(
+                                        "%s Recovery code expired, so generating new one and reseting resend count.",
+                                        LOG_OPERATION);
+                                recoveryCodeMap.put(CodeConstants.CODE.toString(), RandomStringUtils.random(8, true, true));
```
Comment (desouzas) [inline store/src/java/com/zimbra/cs/service/mail/RecoverAccount.java]: Do we want a constant or comment for that `8`?

File: store/docs/soap.txt [hunk 9]
```diff
@@ -4788,14 +4789,15 @@ reset: reads no input parameter. 'reset' will clear all ldap attributes related
...
+recory_email_address = verified recovery email address for getRecoveryAccount operation
```
Comment (desouzas) [inline store/docs/soap.txt]: `recory`

**Comment by chinmay15** [comment]
Discussed with @shrikant-prasad and @Prashantsurana. They suggested some changes. Will work on it and update PR again.

**Comment by chinmay15** [comment]
Continuing on 
https://github.com/Zimbra/zm-mailbox/pull/652

**Pull request closed without merging**